### DISCUSSION TITLE
Throughput Performance on Gateway-to-Gateway is very low after inserting a rule that accepts packets with a matching IPsec policy in the POSTROUTING chain

### DISCUSSION THREAD

**Discussion by shockingchen** [strongswan/strongswan#1944] [Q&A] [answered]
**System (please complete the following information):**
 - OS: [e.g. Ubuntu 18]
 - Kernel version (if applicable): [5.10]
 - strongSwan version(s): [5.9.2]
 - Tested/confirmed with the latest version: [yes]

Gateway-to-Gateway，tunnel mode and esp ；

1.when I did not add any rule in the POSTROUTING 
server
iperf3 -s -p 8000
client
iperf3 -c serIP -u -p 8000 -t 60 -f m -b5000M -P 8 -l 512
result
350 Mbits/sec

2.when I add rules below:
iptables -t nat -A POSTROUTING -o eno0 -j MASQUERADE
iptables -t nat -I POSTROUTING -m policy --pol ipsec --dir out -j ACCEPT

server
iperf3 -s -p 8000
client
iperf3 -c serIP -u -p 8000 -t 60 -f m -b5000M -P 8 -l 512
result
70 Mbits/sec

why the Throughput is so slow  after insert a rule that accepts packets with a matching IPsec policy before any NAT rule in the POSTROUTING chain?
anyone can help to have a look about this?

thank you.

**Comment by tobiasbrunner** [comment]
> -b5000M -P 8

Hm, not sure that makes sense (5000 Mbit/s * 8 = 40 Gbit/s). What's the link speed?

Also, what traffic selectors did you negotiate? What are the endpoints of the packets generated by iperf3?

**Comment by shockingchen** [comment]
ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.10, Linux 5.10.9-rt24-svn2061, aarch64):
  uptime: 22 seconds, since Jan 29 00:23:53 2018
  malloc: sbrk 2568192, mmap 0, used 613600, free 1954592
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 2
  loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf gmp curve25519 xcbc cmac hmac kdf gcm drbg attr kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
Listening IP addresses:
  192.168.0.1
  10.1.0.1
Connections:
        vpn1:  192.168.0.1...192.168.0.2  IKEv2
        vpn1:   local:  [192.168.0.1] uses pre-shared key authentication
        vpn1:   remote: [192.168.0.2] uses pre-shared key authentication
        vpn1:   child:  10.1.0.0/24 === 10.2.0.0/24 TUNNEL
Security Associations (1 up, 0 connecting):
        vpn1[1]: ESTABLISHED 22 seconds ago, 192.168.0.1[192.168.0.1]...192.168.0.2[192.168.0.2]
        vpn1[1]: IKEv2 SPIs: 531844ceabccdeae_i* 764241b714a87f62_r, pre-shared key reauthentication in 51 minutes
        vpn1[1]: IKE proposal: DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
        vpn1{1}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: cd3f4672_i c718dcc2_o
        vpn1{1}:  DES_CBC/HMAC_SHA1_96, 0 bytes_i, 0 bytes_o, rekeying in 15 minutes
        vpn1{1}:   10.1.0.0/24 === 10.2.0.0/24

change the iperf command
client1（10.2.0.12）
iperf3 -s
client2 (10.1.0.10)
iperf3 -c 10.2.0.12 -P 10 -i 1 -t 10
result
100 Mbits/sec

the both client link is 1Gbps
the throughput is just 100 Mbits/sec?how can I improve it?

![image](https://github.com/strongswan/strongswan/assets/49074914/323cc670-7934-4c05-99eb-9d3ad1ecbf77)

**Reply by tobiasbrunner** [reply]
> vpn1{1}: DES_CBC/HMAC_SHA1_96

Ahem, really? Even without hardware that can accelerate AES (e.g. AES-NI), DES will be a lot slower. And the combination of CBC mode with an HMAC is also slower than using an AEAD algorithm like AES-GCM (in particular on hardware with e.g. PCLMULQDQ instruction) or ChaCha20/Poly1305.

**Comment by shockingchen** [comment]
Status of IKE charon daemon (strongSwan 5.8.4, Linux 4.19.232-rt104-g7116c73-dirty, aarch64):
  uptime: 5 seconds, since Jan 01 00:09:57 2000
  malloc: sbrk 2691072, mmap 0, used 612176, free 2078896
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 2
  loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp curve25519 xcbc cmac hmac drbg attr kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
Listening IP addresses:
  192.168.0.2
  10.2.0.1
Connections:
        vpn1:  192.168.0.2...192.168.0.1  IKEv2
        vpn1:   local:  [192.168.0.2] uses pre-shared key authentication
        vpn1:   remote: [192.168.0.1] uses pre-shared key authentication
        vpn1:   child:  10.2.0.0/24 === 10.1.0.0/24 TUNNEL
Security Associations (1 up, 0 connecting):
        vpn1[1]: ESTABLISHED 4 seconds ago, 192.168.0.2[192.168.0.2]...192.168.0.1[192.168.0.1]
        vpn1[1]: IKEv2 SPIs: 5c492b85fb0de26b_i* 47866b9df5ab62bf_r, pre-shared key reauthentication in 55 minutes
        vpn1[1]: IKE proposal: AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
        vpn1{1}:  INSTALLED, TUNNEL, reqid 1, ESP SPIs: ce95aa74_i c6ae16ad_o
        vpn1{1}:  AES_CBC_128/HMAC_MD5_96, 0 bytes_i, 0 bytes_o, rekeying in 16 minutes
        vpn1{1}:   10.2.0.0/24 === 10.1.0.0/24

conn vpn1
        keyexchange=ikev2
        left=192.168.0.2
        leftsubnet=10.2.0.0/24
        leftid=192.168.0.2
        right=192.168.0.1
        rightsubnet=10.1.0.0/24
        rightid=192.168.0.1
        ike=aes-md5-modp1024
        ikelifetime=3600s
        esp=aes-md5-modp1024
        lifetime=3600s
        auto=start

thank you for your reply
 I changed the DES to AES,the throughput can be up to 200 Mbits/sec. Is it the best performance?
 How can I config the ike and esp for the best performance ,How much Mbits/sec it can be reached then？

thank you

**Reply by tobiasbrunner** [reply, marked as answer]
Why would you use MD5 and not SHA-1, which is usually faster, or even AES-GCM?  Anyway, difficult to say as it depends on a lot of things, kernel version, kernel config, loaded modules, hardware, hardware acceleration, drivers, driver config etc.

**Reply by shockingchen** [reply]
thank you very much
